File: production/client/src/main/scala/kpn/client/filter/TimestampFilter.scala
// Migrated to Angular: timestamp-filter.ts
package kpn.client.filter

import japgolly.scalajs.react.Callback
import kpn.shared.TimeInfo
import kpn.shared.Timestamp

class TimestampFilter[T](
  kind: TimeFilterKind.Value,
  fun: (T) => Timestamp,
  timeInfo: TimeInfo,
  selectAll: Callback,
  selectLastWeek: Callback,
  selectLastMonth: Callback,
  selectLastYear: Callback,
  selectOlder: Callback
) extends Filter[T]("lastUpdated") {

  def passes(element: T): Boolean = {
    kind match {
      case TimeFilterKind.ALL => true
      case TimeFilterKind.LAST_WEEK => fun(element) >= timeInfo.lastWeekStart
      case TimeFilterKind.LAST_MONTH => fun(element) >= timeInfo.lastMonthStart
      case TimeFilterKind.LAST_YEAR => fun(element) >= timeInfo.lastYearStart
      case TimeFilterKind.OLDER => fun(element) < timeInfo.lastYearStart
    }
  }

  def filterOptions(allFilters: Filters[T], elements: Seq[T]): Option[FilterOptionGroup] = {
    if (elements.isEmpty) {
      None
    } else {
      val filteredElements = allFilters.filterExcept(elements, this)

      val allCount = filteredElements.size
      val lastWeekCount = filteredElements.count(e => fun(e) >= timeInfo.lastWeekStart)
      val lastMonthCount = filteredElements.count(e => fun(e) >= timeInfo.lastMonthStart)
      val lastYearCount = filteredElements.count(e => fun(e) >= timeInfo.lastYearStart)
      val olderCount = filteredElements.count(e => fun(e) < timeInfo.lastYearStart)

      val enableAll = !(lastWeekCount == allCount || lastMonthCount == allCount || lastYearCount == allCount || olderCount == allCount)

      val all = FilterOption("all", allCount, kind == TimeFilterKind.ALL, selectAll)
      val lastWeek = FilterOption("lastWeek", lastWeekCount, kind == TimeFilterKind.LAST_WEEK, selectLastWeek)
      val lastMonth = FilterOption("lastMonth", lastMonthCount, kind == TimeFilterKind.LAST_MONTH, selectLastMonth)
      val lastYear = FilterOption("lastYear", lastYearCount, kind == TimeFilterKind.LAST_YEAR, selectLastYear)
      val older = FilterOption("older", olderCount, kind == TimeFilterKind.OLDER, selectOlder)

      Some(FilterOptionGroup(name, all, lastWeek, lastMonth, lastYear, older))
    }
  }
}

File: client/src/app/kpn/filter/timestamp-filter.ts
import { TimeInfo } from '@api/common/time-info';
import { Timestamp } from '@api/custom/timestamp';
import { Filter } from './filter';
import { FilterOption } from './filter-option';
import { FilterOptionGroup } from './filter-option-group';
import { Filters } from './filters';
import { TimestampFilterKind } from './timestamp-filter-kind';

export class TimestampFilter<T> extends Filter<T> {
  constructor(
    private readonly kind: TimestampFilterKind,
    private readonly getter: (arg: T) => Timestamp,
    private readonly timeInfo: TimeInfo,
    private readonly selectAll: () => void,
    private readonly selectLastWeek: () => void,
    private readonly selectLastMonth: () => void,
    private readonly selectLastYear: () => void,
    private readonly selectOlder: () => void
  ) {
    super('lastUpdated');
  }

  passes(element: T): boolean {
    if (this.isAll()) {
      return true;
    }
    if (this.isLastWeek()) {
      return this.testLastWeek(element);
    }
    if (this.isLastMonth()) {
      return this.testLastMonth(element);
    }
    if (this.isLastYear()) {
      return this.testLastYear(element);
    }
    if (this.isOlder()) {
      return this.testOlder(element);
    }
    return false;
  }

  filterOptions(allFilters: Filters<T>, elements: T[]): FilterOptionGroup {
    if (elements.length === 0) {
      return null;
    }

    const filteredElements = allFilters.filterExcept(elements, this);

    const allCount = filteredElements.length;
    const lastWeekCount = filteredElements.filter((e) =>
      this.testLastWeek(e)
    ).length;
    const lastMonthCount = filteredElements.filter((e) =>
      this.testLastMonth(e)
    ).length;
    const lastYearCount = filteredElements.filter((e) =>
      this.testLastYear(e)
    ).length;
    const olderCount = filteredElements.filter((e) => this.testOlder(e)).length;

    const all = new FilterOption('all', allCount, this.isAll(), this.selectAll);
    const lastWeek = new FilterOption(
      'lastWeek',
      lastWeekCount,
      this.isLastWeek(),
      this.selectLastWeek
    );
    const lastMonth = new FilterOption(
      'lastMonth',
      lastMonthCount,
      this.isLastMonth(),
      this.selectLastMonth
    );
    const lastYear = new FilterOption(
      'lastYear',
      lastYearCount,
      this.isLastYear(),
      this.selectLastYear
    );
    const older = new FilterOption(
      'older',
      olderCount,
      this.isOlder(),
      this.selectOlder
    );

    return new FilterOptionGroup(
      this.name,
      all,
      lastWeek,
      lastMonth,
      lastYear,
      older
    );
  }

  private sameAsOrYoungerThan(element: T, timestamp: Timestamp): boolean {
    return this.getter(element) >= timestamp;
  }

  private isBetween(element: T, from: Timestamp, to: Timestamp): boolean {
    const value = this.getter(element);
    return value > from && value <= to;
  }

  private olderThan(element: T, timestamp: Timestamp): boolean {
    return this.getter(element) < timestamp;
  }

  private isAll(): boolean {
    return this.kind === TimestampFilterKind.all;
  }

  private isLastWeek(): boolean {
    return this.kind === TimestampFilterKind.lastWeek;
  }

  private isLastMonth(): boolean {
    return this.kind === TimestampFilterKind.lastMonth;
  }

  private isLastYear(): boolean {
    return this.kind === TimestampFilterKind.lastYear;
  }

  private isOlder(): boolean {
    return this.kind === TimestampFilterKind.older;
  }

  private testLastWeek(element: T): boolean {
    return this.sameAsOrYoungerThan(element, this.timeInfo.lastWeekStart);
  }

  private testLastMonth(element: T): boolean {
    return this.isBetween(
      element,
      this.timeInfo.lastMonthStart,
      this.timeInfo.lastWeekStart
    );
  }

  private testLastYear(element: T): boolean {
    return this.isBetween(
      element,
      this.timeInfo.lastYearStart,
      this.timeInfo.lastMonthStart
    );
  }

  private testOlder(element: T): boolean {
    return this.olderThan(element, this.timeInfo.lastYearStart);
  }
}
